models/src/main/scala/coop/rchain/models/SortedParMap.scala

```diff
@@ -1,59 +1,63 @@
 package coop.rchain.models
 
 import coop.rchain.models.rholang.sorter.Sortable
-import coop.rchain.models.rholang.sorter.ordering._
+
 import monix.eval.Coeval
 
 import scala.collection.GenTraversableOnce
-import scala.collection.immutable.HashMap
+import scala.collection.immutable.TreeMap
 
-final class SortedParMap private (ps: Map[Par, Par]) extends Iterable[(Par, Par)] {
+import coop.rchain.models.rholang.sorter.ScoredTerm.{ordering => ScoredTermOrdering}
 
-  // TODO: Merge `sortedList` and `sortedMap` into one VectorMap once available
-  lazy val sortedList: List[(Par, Par)]         = ps.sort
-  private lazy val sortedMap: HashMap[Par, Par] = HashMap(sortedList: _*)
+final case class SortedParMap(ps: TreeMap[Par, Par]) extends Iterable[(Par, Par)] {
 
-  def +(kv: (Par, Par)): SortedParMap = SortedParMap(sortedMap + kv)
+  def +(kv: (Par, Par)): SortedParMap = SortedParMap(ps + (sort(kv._1) -> sort(kv._2)))
 
-  def ++(kvs: GenTraversableOnce[(Par, Par)]): SortedParMap = SortedParMap(sortedMap ++ kvs)
+  def ++(kvs: GenTraversableOnce[(Par, Par)]): SortedParMap =
+    SortedParMap(ps ++ kvs.toStream.map(kv => sort(kv._1) -> sort(kv._2)))
 
-  def -(key: Par): SortedParMap = SortedParMap(sortedMap - sort(key))
+  def -(key: Par): SortedParMap = SortedParMap(ps - sort(key))
 
   def --(keys: GenTraversableOnce[Par]): SortedParMap =
-    SortedParMap(keys.foldLeft(sortedMap) { (map, kv) =>
-      map - sort(kv)
+    SortedParMap(keys.foldLeft(ps) { (map, key) =>
+      map - sort(key)
     })
 
-  def apply(par: Par): Par = sortedMap(sort(par))
-
-  def contains(par: Par): Boolean = sortedMap.contains(sort(par))
```

Comment:
as mentioned in discord, lookups would be more efficient if we kept both a hashmap and a treemap, and did the lookups inside the hashMap